src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java

```diff
@@ -75,16 +103,166 @@ private static FileSystem createFileSystem(URI uri, Configuration conf)
+            throws IOException
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    private static FileSystemKey createFileSystemKey(URI uri, UserGroupInformation userGroupInformation, long unique)
+    {
+        String scheme = uri.getScheme() == null ? "" : uri.getScheme().toLowerCase(ENGLISH);
+        String authority = uri.getAuthority() == null ? "" : uri.getAuthority().toLowerCase(ENGLISH);
+        String realUser;
+        String proxyUser;
+        AuthenticationMethod authenticationMethod = userGroupInformation.getAuthenticationMethod();
+        switch (authenticationMethod) {
+            case SIMPLE:
+            case KERBEROS:
+                realUser = userGroupInformation.getUserName();
+                proxyUser = null;
+                break;
+            case PROXY:
+                realUser = userGroupInformation.getRealUser().getUserName();
+                proxyUser = userGroupInformation.getUserName();
+                break;
+            default:
+                throw new IllegalArgumentException("Unsupported authentication method: " + authenticationMethod);
+        }
+        return new FileSystemKey(scheme, authority, unique, realUser, proxyUser);
+    }
+
+    private static Set<?> getPrivateCredentials(UserGroupInformation userGroupInformation)
+    {
+        AuthenticationMethod authenticationMethod = userGroupInformation.getAuthenticationMethod();
+        switch (authenticationMethod) {
+            case SIMPLE:
+                return ImmutableSet.of();
+            case KERBEROS:
+                return ImmutableSet.copyOf(getSubject(userGroupInformation).getPrivateCredentials());
+            case PROXY:
+                return getPrivateCredentials(userGroupInformation.getRealUser());
+            default:
+                throw new IllegalArgumentException("Unsupported authentication method: " + authenticationMethod);
+        }
+    }
+
+    private static class FileSystemKey
+    {
+        private final String scheme;
+        private final String authority;
+        private final long unique;
+        private final String realUser;
+        private final String proxyUser;
+
+        private FileSystemKey(String scheme, String authority, long unique, String realUser, String proxyUser)
+        {
+            this.scheme = requireNonNull(scheme, "scheme is null");
+            this.authority = requireNonNull(authority, "authority is null");
+            this.unique = unique;
+            this.realUser = requireNonNull(realUser, "realUser");
+            this.proxyUser = proxyUser;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            FileSystemKey that = (FileSystemKey) o;
+
+            return Objects.equals(this.scheme, that.scheme) &&
```

Comment:
Nit: all the `this.` are not required (and the IntelliJ generator doesn't put them in for the default template)